common/gic.h

```diff
@@ -0,0 +1,68 @@
+#ifndef _GIC_H
+#define _GIC_H
+
+// These are only accessible from arm11 core...
+
+#define MPCORE_PRIV_REG_BASE	(0x17E00000)
+#define GIC_CPU_REGS_BASE	(MPCORE_PRIV_REG_BASE + 0x0100)
+#define GIC_DIST_REGS_BASE	(MPCORE_PRIV_REG_BASE + 0x1000)
+
+/*
+ * Names/values via:
+ * https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/irqchip/arm-gic.h
+ */
+
+#define GIC_CPU_CTRL		(GIC_CPU_REGS_BASE + 0x00)
+#define GIC_DIST_ENABLE_SET	(GIC_DIST_REGS_BASE + 0x100)
+#define GIC_DIST_ENABLE_CLEAR	(GIC_DIST_REGS_BASE + 0x180)
+#define GIC_DIST_PENDING_CLEAR	(GIC_DIST_REGS_BASE + 0x280)
+#define GIC_DIST_PRI		(GIC_DIST_REGS_BASE + 0x400)
+#define GIC_DIST_TARGET		(GIC_DIST_REGS_BASE + 0x800)
+
+#define BIT(n) (1 << (n))
+
+#ifndef __ASSEMBLY__
+
+static inline void gic_enable_control(void) {
+	*(volatile int*)GIC_CPU_CTRL = 1;
+}
+
+static inline void gic_disable_all(void)
+{
+	volatile int *gic_enable_clear = (volatile int*)GIC_DIST_ENABLE_CLEAR;
+	gic_enable_clear[0] = -1;
+	gic_enable_clear[1] = -1;
+	gic_enable_clear[2] = -1;
+	gic_enable_clear[3] = -1;
+}
+
+static inline void gic_x_interrupt(volatile int* addr, int id)
+{
+	addr[0] = BIT(id);
+	addr[1] = BIT(id % 32);
+	addr[2] = BIT(id % 32);
+	addr[3] = BIT(id % 32);
+}
+
+static inline void gic_clear_interrupt(int id)
+{
+	gic_x_interrupt((volatile int*)GIC_DIST_PENDING_CLEAR, id);
+}
+
+static inline void gic_set_prio(int id, int prio)
+{
+	*(volatile int*)(GIC_DIST_PRI + id) = prio;
+}
+
+static inline void gic_set_target(int id, int core)
+{
+	*(volatile int*)(GIC_DIST_TARGET + id) = core + 1;
+}
```

Comment:
@profi200 I'm not sure I interpreted [this line](https://github.com/profi200/open_agb_firm/blob/1fcf024520571c9e8732a5aaaf1bbd7661cba423/source/arm11/hardware/pdn.c#L70) correctly. How many interrupts can we have? The members of `GicDist` make it look like there's 8*32 possible, but then in your implementation only the first 4*32 get cleared?

Reply:
This older GIC (not officially called GIC) supports 256 interrupt lines max. The 3DS uses a maximum of 128.

https://www.3dbrew.org/wiki/ARM11_Interrupts

Reply:
> not officially called GIC

heh, perhaps better than the acronym for Distributed Interrupt Controller?  (Thanks again for very helpful links, how do you get there from the main page?)

Reply:
ah, I see the wiki has "what links here" links on the left, so I can trace back to the homepage.  I'll add some of these links as comments in the code.

Reply:
It's a little hidden but i found it by searching for interrupts.

Reply:
I guess I'm still curious though about the comment that says "target core 0" but then assigns a value of 1.  Is the "target" the CPUID plus one?

Reply:
It's a bitmap. One bit for each CPU core. See here: https://developer.arm.com/documentation/ddi0360/f/mpcore-distributed-interrupt-controller/interrupt-distributor-registers/interrupt-cpu-targets-registers--0x800-0x8fc?lang=en

Reply:
thanks, changed from `core + 1` to `BIT(core)`.